notebooks/official/ray_on_vertex_ai/train/model_training_formalization_with_ray_and_kfp.ipynb

```diff
@@ -0,0 +1,1765 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "J7gHWrN3kur-"
+      },
+      "outputs": [],
+      "source": [
+        "@dsl.component(\n",
+        "    base_image=VERTEX_RAY_IMAGE_URI,\n",
+        "    packages_to_install=[\n",
+        "        \"google-cloud-aiplatform[ray]\",\n",
+        "        \"ray[data]==2.9.3\",\n",
+        "        \"ray[data]==2.9.3\",\n",
+        "        \"ray[train]==2.9.3\",\n",
+        "        \"kfp\",\n",
+        "        \"google-cloud-pipeline-components\",\n",
+        "        \"etils\",\n",
+        "        \"importlib_resources\",\n",
+        "    ],\n",
+        ")\n",
+        "def GetMetricsOp(\n",
+        "    training_dir: str, training_run_id: str, metrics: Output[Metrics]\n",
+        ") -> NamedTuple(\"outputs\", threshold_metric=float):\n",
+        "    \"\"\"Get evaluation metrics from training run.\"\"\"\n",
+        "\n",
+        "    import logging\n",
+        "\n",
+        "    from etils import epath\n",
+        "    from ray.tune import ExperimentAnalysis\n",
+        "\n",
+        "    eval_metrics = \"error\"\n",
+        "    threshold_metric = \"valid-error\"\n",
+        "\n",
+        "    # Configure logging\n",
+        "    logging.basicConfig(\n",
+        "        level=logging.INFO, format=\"%(asctime)s - %(levelname)s - %(message)s\"\n",
+        "    )\n",
+        "\n",
+        "    # restore training result and log metrics\n",
+        "    logging.info(\"Restoring training result.\")\n",
+        "    training_dir = training_dir.replace(\"gs://\", \"/gcs/\")\n",
+        "    training_path = epath.Path(training_dir)\n",
+        "    training_id_path = training_path / training_run_id\n",
+        "    experiment_analysis = ExperimentAnalysis(training_id_path)\n",
+        "    results = experiment_analysis.results_df.reset_index().to_dict(orient=\"records\")\n",
+        "\n",
+        "    # log metrics and store threshold metrics\n",
+        "    logging.info(\"Logging metrics and store threshold metrics.\")\n",
+        "    for result in results:\n",
+        "        for key in result.keys():\n",
+        "            if eval_metrics in key:\n",
+        "                metrics.log_metric(key, round(result[key], 5))\n",
+        "\n",
+        "    threshold_metric = results[0].get(threshold_metric)\n",
+        "\n",
+        "    component_outputs = NamedTuple(\"outputs\", threshold_metric=float)\n",
+        "    return component_outputs(round(threshold_metric, 5))"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "1-Ccx4uLDz4N"
+      },
+      "source": [
+        "#### Build Model Versioning component\n",
+        "\n",
+        "The `UploadRayModelOp` component collects the best training checkpoint and it stores the model to Vertex AI Model Registry with/without explainability configuration."
```

Comment:
- don't need "it" and issue with "/"
- recommend changing to: "The `UploadRayModelOp` component collects the best training checkpoint and stores the model to Vertex AI Model Registry with or without explainability configuration."